action.yml

```diff
@@ -20,10 +20,52 @@ branding:
 runs:
   using: composite
   steps:
+    # Set up a non-EOL, cibuildwheel & pipx supported Python version
+    - name: Find python interpreter
+      id: python
+      run: |
+        from pathlib import Path
+        import os
+        import platform
+        import sys
+
+        SUPPORTED_VERSIONS=("3.10", "3.9", "3.8", "3.7", "3.6")
+        EXT = "${{ runner.os == 'Windows' && '.exe' || '3' }}"
+        PREFIX = "${{ runner.os != 'Windows' && 'bin/' || '' }}"
+        ARCH_MAP = {"AMD64": "x64", "x86_64": "x64"}
+
+        arch = ARCH_MAP.get(platform.machine(), None)
+        base_path = Path(os.environ.get("RUNNER_TOOL_CACHE", "not_existing")) / "Python"
+        if arch and base_path.exists():
+            for version in SUPPORTED_VERSIONS:
+                pattern = "{}.*/{}/{}python{}".format(version, arch, PREFIX, EXT)
+                matches = list(base_path.glob(pattern))
+                if matches:
+                    exact_version = matches[0].relative_to(base_path).parts[0]
+                    print("Using cached CPython {} at {}".format(exact_version, matches[0]))
+                    print("::set-output name=interpreter-path::{}".format(matches[0]))
+                    exit(0)
+        # fallback
+        print("::warning ::Couldn't find a suitable python version in the toolcache, fallback to current interpreter")
+        current_version = ".".join(map(str, sys.version_info[:2]))
+        if current_version in SUPPORTED_VERSIONS:
+            exact_version = ".".join(map(str, sys.version_info[:3]))
+            print("Using Python {} at {}".format(exact_version, sys.executable))
+            print("::set-output name=interpreter-path::{}".format(sys.executable))
+            exit(0)
+        print(
+            "::error ::The current interpreter is Python {}, cibuildwheel requires one of {}.".format(
+                current_version, ", ".join(SUPPORTED_VERSIONS)
+            ),
+            "You can use actions/setup-python to install a supported version."
+        )
+        exit(1)
+      shell: "${{ runner.os == 'Windows' && 'python' || 'python3' }} {0}"
```

Comment:
This could be reworked as a helper script with its own tests.
Syntax is Python 3.4+
Unfortunately, setup-python messes with the environment for other steps & I'd rather avoid side-effects for the rest of users workflow when using the cibuildwheel GitHub Action.
It should allow self-hosted runners to continue working out-of-the-box.

Reply:
Can't this just be `python`? https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#example-running-a-python-script

Reply:
This could be an action, we could put it in scikit-build/latest-python or something like that.

Reply:
> or does that get Python 2 somewhere (like on macOS)?

Indeed... https://github.com/mayeut/cibuildwheel/actions/runs/2373174130

| runner | python |
|---|---|
|ubuntu-18.04|2.7.17|
|ubuntu-20.04|3.8.10|
|ubuntu-22.04|3.10.4|
|windows-2019|3.7.9|
|windows-2022|3.9.12|
|macos-10.15|2.7.18|
|macos-11|2.7.18|
|macos-12|2.7.18|

> If not, why not python3 {0}?

It works https://github.com/mayeut/cibuildwheel/actions/runs/2373224962. Good to know they've added python3 on windows !

| runner | python |
|---|---|
|ubuntu-18.04|3.6.9|
|ubuntu-20.04|3.8.10|
|ubuntu-22.04|3.10.4|
|windows-2019|3.7.9|
|windows-2022|3.9.12|
|macos-10.15|3.9.12|
|macos-11|3.9.12|
|macos-12|3.9.12|

> This could be an action, we could put it in scikit-build/latest-python or something like that.

Yes, we could. I'll try to get this into [actions/setup-python](https://github.com/actions/setup-python). (option not to mess with the environment, together with an `python-path` output - already got a PR for that last one)

Reply:
> I'll try to get this into [actions/setup-python](https://github.com/actions/setup-python).

PR created at https://github.com/actions/setup-python/pull/411

I'll wait for feedback there to see which direction this PR will take (marking as draft for now)

Reply:
It seems https://github.com/actions/setup-python/pull/411 will be merged in some form or another